### DISCUSSION TITLE
GraphSAINT test

### DISCUSSION THREAD

**Discussion by asd10asd** [pyg-team/pytorch_geometric#2545] [Q&A] [answered]
Hi,

I have a large graph(about 50,000 nodes,10,000,000 edges) and try to train a GNN model . But it seems that the graph is too large to train. I try to use GraphSAINT and read the example graph_saint.py. But when test on the whole graph, it can't work beacuse of limited memory. How can I test the whole graph? Or maybe there is a more convenient way to train and test on my data?

Thanks!

**Comment by rusty1s** [comment, marked as answer]
50k nodes isn't that large, so I suggest to look into our [memory-efficient aggregation options](https://pytorch-geometric.readthedocs.io/en/latest/notes/sparse_tensor.html).

**Reply by asd10asd** [reply]
Thank you for your reply. I have changed my code to use SparseTensor to save memory. But now I ran into another problem. My initialized feature matrix X is an identity matrix. This matrix is 50000* 50000 and need about 12G memory. Can I also use sparse matrix multiplication for X？

**Reply by rusty1s** [reply]
Yes, this is possible for some, but not all GNN operators. I guess a more elegant solution is to utilize `torch.nn.Embedding` for learning node features rather than relying on identity matrices (which basically pushes the task of learning node features to the first GNN op).

**Reply by asd10asd** [reply]
Thank you very much. Your answer helped me very well. I have solved this problem now.
